### ISSUE TITLE
Multi company issue

### ISSUE THREAD

**Issue by aitorbouzas** [WilldooIT/Pentaho-reports-for-OpenERP#256]
Hello,

I have an installation of Odoo 8 with Pentaho Reports.
I'm having an issue where two companies are throwing reports with no data (only the labels I have defined in the report). When we switch to the other company, sometimes deleting the cookies solves the problem, but sometimes it doesn't.

This is the company tree:
1. Coolabora
     1.a Conecta
     1.b External

Now it seems that deleting the cookies won't make it work again... I don't know what to do.

**Comment by richard-willdooit** [comment]
It is an object based report or an sql  query?  If object, then visibility of data will depend on the currently selected company of the user executing the report.

Also, there may be a problem, depending on what version of the modules you are using.  To get around multi-user concurrency, the system creates a duplicate of the user to run the report.  If that user record is not deleted, then there is a possibility when you run another report, that the existing duplicate user (and therefore the company id that WAS selected) is the one it runs for.

Again, this is for object reports only.

**Comment by aitorbouzas** [comment]
I see, so I could solve this by making it sql query instead of object based or by upgrading the module?
I'll give it a try and come back.

Thanks

**Comment by aitorbouzas** [comment]
I have another question related to the answer you gave me @richard-willowit 

Does this means that if user 1 in company 1 runs a report just before user 2 in company 2 does it, the pentaho will (maybe) try to run the report with the user 1 company 1 stored session?

Is there anyway to surpass this? 
I tried to do it by creating the reports in sql based query but when another user ran another report, I couldn't do it anymore (I suppose because it happened what you told me about the stored sessions...)

Thanks

**Comment by richard-willdooit** [comment]
@Softdil Not at all.  When the duplicate "pentaho" user code was in place, it used one per base user.  So user 1 and user 2 would never be a conflict, just if user 1 ran and the session died, and then user 1 ran again in another company (although, as I hinted, I have not looked to see if this is your problem, it is just a pointer).

**Comment by richard-willdooit** [comment]
@Softdil Running the reports as isolated separate users should never be an issue, sql or object based.  The only time I get worried is if the code never cleanly cleaned up the temporary user id.  And that MAY be a problem if the same user runs reports sequentially, and PROBABLY is a problem if the same user runs reports in parallel....

**Comment by aitorbouzas** [comment]
Ok, so the problem is when an user runs several reports in different companies... That's practically our case and I think, as you say, that the session is beeing stored somewhere with the old company id, so the report shows blank...
Is there any workaround I could try with this?

**Comment by richard-willdooit** [comment]
Firstly, ascertain whether the "pentaho" user record (the copy of the res.users) is in fact being left behind after running or whether it is cleaning up correctly when the report is executed.

**Comment by aitorbouzas** [comment]
How can I check that? 
Now i'm having a similar problem (maybe it's related). When I print a report of a contract with start data and end date it prints correct the header and two sub-reports I made. The problem comes as I change these dates to check the result of those two other sub-reports... The header shows ok, but the sub reports are showing the same as before (the header and main report is object based, and the sub-reports are SQL)

Waiting your answers.

**Comment by aitorbouzas** [comment]
Please, I need to know where to check this...

**Comment by richard-willdooit** [comment]
@softdil in res.users.  If the record is there in res.users or not after running the report.

Also, have you checked the report is not caching?  Check the following in the report designer:

Master Report --> Attributes --> query --> data-cache --> false
Master Report --> Attributes --> pentaho --> report-cache --> false

**Comment by aitorbouzas** [comment]
I was able to reproduce the error checking the table res.users.
This is the problem:

When user 1 is able to print into company 1, a record is created correctly and it's dropped after the report is done... but when it tries to print another record in company 2, it creates an user with company 1 so the report doesn't print correctly because it's not allowed...

**Comment by aitorbouzas** [comment]
I see that in the code there's no line where the company_id is passed to the temp user... Maybe we can fix this to take the actual company of this user?

**Comment by aitorbouzas** [comment]
I think I fixed this by updating my module code... It was not the last, and it was still using the "find_temp_user" (or so) method... I'll come back if the problem repeats.
Thanks for your help
